float32 phi
float32 theta
float32 psi 